src/keystore.cpp

```diff
@@ -157,7 +157,7 @@ bool CBasicKeyStore::AddSaplingFullViewingKey(
     auto ivk = fvk.in_viewing_key();
     mapSaplingFullViewingKeys[ivk] = fvk;
 
-    return AddSaplingIncomingViewingKey(ivk, defaultAddr);
+    return CBasicKeyStore::AddSaplingIncomingViewingKey(ivk, defaultAddr);
```

Comment:
I believe the deadlock was caused by the fact that `CWalletDB::LoadWallet` when loading a Sapling key calls `LoadSaplingZKey`, which calls `CCryptoKeyStore::AddSaplingSpendingKey`, which ends up calling here. If we call `AddSaplingIncomingViewingKey` here then the wallet ends up trying to write the result out, which likely causes a deadlock in the underlying database.

The effect of this change is that when creating a new Sapling key, the `defaultAddr -> ivk` map entry is not persisted to disk (while other `addr -> ivk` maps are, because they call `AddSaplingIncomingViewingKey` directly). However, this is fine, because we can always recover the default address from the main key entry (either unencrypted from the extended spending key, or encrypted via the extended full viewing key).